From 2155f4118e2e49f48d52a1366c51021b53e1f0c0 Mon Sep 17 00:00:00 2001
From: Thomas Langer <thomas.langer@intel.com>
Date: Thu, 10 Oct 2019 15:54:23 +0200
Subject: [PATCH] PONRTSYS-5112: Try harder to find a relocation offset

Add some brute-force to make sure to find some offset
which does not overlap the old kernel location and
will not go behind CONFIG_RANDOMIZE_BASE_MAX_OFFSET.
---
 arch/mips/kernel/relocate.c | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/arch/mips/kernel/relocate.c b/arch/mips/kernel/relocate.c
index b4a7c303019b..14d5e0d4f36f 100644
--- a/arch/mips/kernel/relocate.c
+++ b/arch/mips/kernel/relocate.c
@@ -248,16 +248,28 @@ static inline void __init *determine_relocation_address(void)
 	unsigned long kernel_length;
 	void *dest = &_text;
 	unsigned long offset;
+	int count = 100;
 
 	if (kaslr_disabled())
 		return dest;
 
 	kernel_length = (long)_end - (long)(&_text);
 
-	offset = get_random_boot() << 16;
-	offset &= (CONFIG_RANDOMIZE_BASE_MAX_OFFSET - 1);
-	if (offset < kernel_length)
-		offset += ALIGN(kernel_length, 0xffff);
+	do {
+		offset = get_random_boot() << 16;
+		offset &= (CONFIG_RANDOMIZE_BASE_MAX_OFFSET - 1);
+		if (offset < kernel_length)
+			offset += ALIGN(kernel_length, 0xffff);
+
+		if (offset + kernel_length <= CONFIG_RANDOMIZE_BASE_MAX_OFFSET)
+			break;
+		pr_debug("Relocation by 0x%lX not possible\n", offset);
+	} while (--count);
+
+	if (!count) {
+		pr_err("Relocation not possible, did not find a working offset\n");
+		offset = 0;
+	}
 
 	return RELOCATED(dest);
 }
